[WIP] Add Element class for manipulating DOM elements in Lua

mike1808 opened the pull request:
some test lua code

```lua
function main(splash)
  assert(splash:go('http://localhost:8000/'))

  local element = splash:select('button')
  element:mouse_click()
  assert(splash:wait(0.0001))

  local title = splash:select('#title')

  return {title:get_styles()["display"], title:node_property('nodeName')}
end
```
for the following html
```html
<!DOCTYPE html>
<html>
<head>
   <title>Index 1</title>
</head>
<body>
<button type="button" onclick="document.getElementById('title').style.display = 'block'">Click me</button>
<h1 id="title" style="display: none">Index 1</h1>
</body>
</html>
```

Youwotma commented on splash/browser_tab.py:
```diff
@@ -1078,6 +1093,65 @@ def log(self, message, min_level=None):
+class HTMLElement(object):
+    DIMENSIONS_JS_FUNC = """
+        (function(element) {
+            var rect = element.getBoundingClientRect();
```
better `getClientRects()[0]`

mike1808 replied:
Why it is better?

Youwotma replied:
Consider this link: ![screenshoot](http://i.imgur.com/hgVDv9x.png)

The blue semi-transparent rectangles are the client rects (it has two), the dotted line the bounding client rect (the rect that contains the client rects).

The top left corner of the bounding client rect is not "on" the element, the top-left of one of the client rects is.

immerrr replied:
That might be a problem if you want to "click" the element. If you want to take a screenshot, you most likely won't be satisfied with just the first client rect.

Youwotma replied:
Yes, exactly, both of the methods exist for a reason. It might also be
worth it to perform a hittest on the coordinates and raise an error if its
not successful. The element might be hidden behind another element or
transparent in that coordinates, so raising an error would be better than
failing silently and possibly causing side effects.
On Jul 7, 2016 12:08, "dennis again" <notifications@github.com> wrote:

> In splash/browser_tab.py
> <https://github.com/scrapinghub/splash/pull/471#discussion_r69851779>:
>
> > @@ -1078,6 +1093,65 @@ def log(self, message, min_level=None):
> >          log.msg(message, system='render')
> >
> >
> > +class ElementsStorage(QObject):
> > +    def __init__(self, parent):
> > +        self.name = str(uuid.uuid1())
> > +        super(ElementsStorage, self).__init__(parent)
> > +
> > +    @pyqtSlot(name="getId", result=str)
> > +    def get_id(self):
> > +        return str(uuid.uuid1())
> > +
> > +
> > +class HTMLElement(object):
> > +    DIMENSIONS_JS_FUNC = """
> > +        (function(element) {
> > +            var rect = element.getBoundingClientRect();
>
> That might be a problem if you want to "click" the element. If you want to
> take a screenshot, you most likely won't be satisfied with just the first
> client rect.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scrapinghub/splash/pull/471/files/37b17cdce31ab27b2b866f6afa13b69f2607c059#r69851779>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AANrvUUUhqQV2uQ65VdjchSgpWpEm0oTks5qTInngaJpZM4I8Dce>
> .
>

immerrr replied:
True, but I still think that for getting "dimensions" the bounding variant is better. For clicking or interacting with it somehow, probably not, but this method is used later on for capturing the elements and like I said, the first client rect is most certainly not what you'd expect when you'd do `element.png()`.

immerrr replied:
As a sidenote, @mike1808, would it be useful to return `nil` from `png` method if the element is not visible?

mike1808 replied:
@immerrr it sounds reasonable.

Youwotma commented on splash/browser_tab.py:
```diff
@@ -1078,6 +1093,65 @@ def log(self, message, min_level=None):
+
+    def node_method(self, method_name):
+        self.assert_element_exists()
+        return self.tab.evaljs("%s[%s]()" % (self.element_js, escape_js(method_name)))
```
Should we use `jsfunc` and return the function, so the user can call it with whatever arguments they want? You might need to bind the element to the context or wrap the function for it to work properly.

mike1808 replied:
You are right. I'll change it.

immerrr commented on splash/html_element.py:
```diff
@@ -0,0 +1,222 @@
+
+    def get_bounds(self):
+        self.assert_element_exists()
+        return self.tab.evaljs("%s.getClientRects()[0]" % self.element_js, result_protection=False)
```
This getClientRects/getBoundingClientRect debate makes me wonder:
- if we want to expose getBoundingClientRect to the users
- if we want to expose getClientRects returning an array to the users
- if we want to capture (png) all client rects or the bounding client rect to the users

mike1808 commented:
@Youwotma @immerrr I've changed the method of getting boundary rectangle for `png` and `jpeg` methods to `getBoundaryRectangle`. Now those methods return `None` if element is not visible.

Also I moved all used Casper.JS methods to separate file and added MIT license to them. 

@Youwotma check please the `node_method` both in Lua and Python. I have 2 different implementation for Python API and Lua.

mike1808 commented on splash/qtrender_lua.py:
```diff
@@ -328,7 +329,7 @@ def __call__(self, *args):
         )
         wrapper_script = get_process_errors_js(expr)
-        res = self.tab.evaljs(wrapper_script)
+        res = self.tab.evaljs(wrapper_script, result_protection=self.js_result_protection)
```
If you turn on `result_protection` some objects cannot be retreived, for example `getBoundaryRect()` returns an instance of [DOMRect](https://developer.mozilla.org/en-US/docs/Web/API/DOMRect) which is ignored in sanitize function.

kmike replied:
I think the way to go is to wrap objects like DOMRect explicitly.

mike1808 replied:
@kmike what do you mean? Should new rules be added to [sanitize](https://github.com/scrapinghub/splash/blob/master/splash/jsutils.py#L12) function? Or it should be written like
```js
var rect = elem. getBoundingClientRect();
return { top: rect.x, right: rect.right, bottom: rect.bottom, left: rect.left };
```

kmike replied:
I think both these options are fine. Adding it to sanitize may be a bit more convenient for users because they won't have to write `{ top: rect.x, right: rect.right, bottom: rect.bottom, left: rect.left }` themselves. A third option is to convert DOMRect to a special DOMRect Lua object - just add a few useful methods to `{ top=rect.x, right=rect.right, bottom=rect.bottom, left=rect.left }` table.

mike1808 commented:
Fixed bugs, rebased code, added tests. Some of them fails because of #488 
This patch fixes that but I didn't want to include it in this PR:
```
From d9b0204eaa74d8836cd9ae66567836b5f737a34a Mon Sep 17 00:00:00 2001
From: Michael Manukyan <arm.localhost@gmail.com>
Date: Tue, 26 Jul 2016 20:39:25 +0400
Subject: [PATCH] store private methods in the instance itself rather than in
 the separate table

---
 splash/lua_modules/extras.lua    |  3 +--
 splash/lua_modules/request.lua   |  3 +--
 splash/lua_modules/response.lua  |  3 +--
 splash/lua_modules/splash.lua    | 22 ++++++++++------------
 splash/lua_modules/wraputils.lua | 16 +++++++++-------
 5 files changed, 22 insertions(+), 25 deletions(-)

diff --git a/splash/lua_modules/extras.lua b/splash/lua_modules/extras.lua
index 61bf6fa..8a0e0f1 100644
--- a/splash/lua_modules/extras.lua
+++ b/splash/lua_modules/extras.lua
@@ -6,12 +6,11 @@
 local wraputils = require("wraputils")
 
 local Extras = {}
-local Extras_private = {}
 wraputils.set_metamethods(Extras)
 
 function Extras._create(py_extras)
   local extras = {}
-  wraputils.wrap_exposed_object(py_extras, extras, Extras, Extras_private, false)
+  wraputils.wrap_exposed_object(py_extras, extras, Extras)
   return extras
 end
 
diff --git a/splash/lua_modules/request.lua b/splash/lua_modules/request.lua
index 79d5783..00caeb8 100644
--- a/splash/lua_modules/request.lua
+++ b/splash/lua_modules/request.lua
@@ -5,7 +5,6 @@ local wraputils = require("wraputils")
 local treat = require("libs/treat")
 
 local Request = wraputils.create_metatable()
-local Request_private = {}
 wraputils.set_metamethods(Request)
 
 function Request._create(py_request)
@@ -16,7 +15,7 @@ function Request._create(py_request)
     method=py_request.method,
   }
 
-  wraputils.wrap_exposed_object(py_request, request, Request, Request_private, false)
+  wraputils.wrap_exposed_object(py_request, request, Request)
   return request
 end
 
diff --git a/splash/lua_modules/response.lua b/splash/lua_modules/response.lua
index 65ac2d8..d8c44d8 100644
--- a/splash/lua_modules/response.lua
+++ b/splash/lua_modules/response.lua
@@ -6,7 +6,6 @@ local treat = require("libs/treat")
 local Request = require("request")
 
 local Response = wraputils.create_metatable()
-local Response_private = {}
 wraputils.set_metamethods(Response)
 
 function Response._create(py_response)
@@ -15,7 +14,7 @@ function Response._create(py_response)
     request=Request._create(py_response.request),
   }
 
-  wraputils.wrap_exposed_object(py_response, response, Response, Response_private, false)
+  wraputils.wrap_exposed_object(py_response, response, Response)
   return response
 end
 
diff --git a/splash/lua_modules/splash.lua b/splash/lua_modules/splash.lua
index 2598ead..933992f 100644
--- a/splash/lua_modules/splash.lua
+++ b/splash/lua_modules/splash.lua
@@ -12,12 +12,11 @@ local repr = require("repr")
 -- Lua wrapper for Splash Python object.
 --
 local Splash = wraputils.create_metatable()
-local Splash_private = {}
 wraputils.set_metamethods(Splash)
 
 function Splash._create(py_splash)
   local splash = { args = py_splash.args }
-  wraputils.wrap_exposed_object(py_splash, splash, Splash, Splash_private, true)
+  wraputils.wrap_exposed_object(py_splash, splash, Splash)
   return splash
 end
 
@@ -26,7 +25,7 @@ end
 -- It is required to handle errors properly.
 --
 function Splash:jsfunc(...)
-  local func = Splash_private.jsfunc(self, ...)
+  local func = self.__private.jsfunc(self, ...)
   return wraputils.unwraps_python_result(func)
 end
 
@@ -37,7 +36,7 @@ function Splash:on_request(cb)
   if type(cb) ~= 'function' then
     error("splash:on_request callback is not a function", 2)
   end
-  Splash_private.on_request(self, function(py_request)
+  self.__private.on_request(self, function(py_request)
     local req = Request._create(py_request)
     return cb(req)
   end)
@@ -47,7 +46,7 @@ function Splash:on_response_headers(cb)
   if type(cb) ~= 'function' then
     error("splash:on_response_headers callback is not a function", 2)
   end
-  Splash_private.on_response_headers(self, function(response)
+  self.__private.on_response_headers(self, function(response)
     local res = Response._create(response)
     return cb(res)
   end)
@@ -57,7 +56,7 @@ function Splash:on_response(cb)
   if type(cb) ~= 'function' then
     error("splash:on_response callback is not a function", 2)
   end
-  Splash_private.on_response(self, function(response)
+  self.__private.on_response(self, function(response)
     local res = Response._create(response)
     return cb(res)
   end)
@@ -68,17 +67,16 @@ end
 -- Timer Lua wrapper
 --
 local Timer = wraputils.create_metatable()
-local Timer_private = {}
 wraputils.set_metamethods(Timer)
 
 function Timer._create(py_timer)
   local timer = {}
-  wraputils.wrap_exposed_object(py_timer, timer, Timer, Timer_private, true)
+  wraputils.wrap_exposed_object(py_timer, timer, Timer)
   return timer
 end
 
 function Splash:call_later(cb, delay)
-  local py_timer = Splash_private.call_later(self, cb, delay)
+  local py_timer = self.__private.call_later(self, cb, delay)
   return Timer._create(py_timer)
 end
 
@@ -97,7 +95,7 @@ function Element._create(py_element)
 end
 
 function Element:node_method(...)
-  local ok, func = Element_private.node_method(self, ...)
+  local ok, func = self.__private.node_method(self, ...)
 
   if not ok then
     return ok, func
@@ -108,7 +106,7 @@ end
 
 
 function Element:node_property(...)
-  local ok, result, is_element = Splash_private.node_property(self, ...)
+  local ok, result, is_element = self.__private.node_property(self, ...)
 
   if not ok then
     return ok, result
@@ -123,7 +121,7 @@ end
 
 
 function Splash:select(selector)
-  local py_element = Splash_private.select(self, selector)
+  local py_element = self.__private.select(self, selector)
   return Element._create(py_element)
 end
 
diff --git a/splash/lua_modules/wraputils.lua b/splash/lua_modules/wraputils.lua
index cd8a631..671b916 100644
--- a/splash/lua_modules/wraputils.lua
+++ b/splash/lua_modules/wraputils.lua
@@ -97,10 +97,12 @@ end
 -- * Lua methods are created for Python methods wrapped in @command.
 -- * Async methods are wrapped with `coroutine.yield`.
 -- * Lua <-> Python error handling is fixed.
--- * Private methods are stored in `private_self`, public methods are
+-- * Private methods are stored in `self.__private`, public methods are
 --   stored in `self`.
 --
-local function setup_commands(py_object, self, private_self)
+local function setup_commands(py_object, self)
+  rawset(self, '__private', {})
+
   -- Create lua_object:<...> methods from py_object methods:
   for key, opts in pairs(py_object.commands) do
     local command = py_object[key]
@@ -121,7 +123,7 @@ local function setup_commands(py_object, self, private_self)
 
     if is_private_name(key) then
       local short_key = string.sub(key, PRIVATE_PREFIX:len() + 1)
-      private_self[short_key] = command
+      self.__private[short_key] = command
     else
       -- avoid custom setter
       rawset(self, key, command)
@@ -133,7 +135,7 @@ end
 --
 -- Handle @lua_property decorators.
 --
-local function setup_property_access(py_object, self, cls)
+local function setup_property_access(py_object, self)
   rawset(self, '__getters', {})
   rawset(self, '__setters', {})
 
@@ -153,10 +155,10 @@ end
 --
 -- Create a Lua wrapper for a Python object.
 --
-local function wrap_exposed_object(py_object, self, cls, private_self)
+local function wrap_exposed_object(py_object, self, cls)
   setmetatable(self, cls)
-  setup_commands(py_object, self, private_self)
-  setup_property_access(py_object, self, cls)
+  setup_commands(py_object, self)
+  setup_property_access(py_object, self)
 end
 
 
-- 
2.7.4 (Apple Git-66)
```

mike1808 commented:
@immerrr @kmike I've added docs and test.

immerrr commented on docs/scripting-element-object.rst:
```diff
@@ -0,0 +1,395 @@
+
+**Returns:** ``ok, exists`` pair. If ``ok`` is nil then error happened during the check;
+``exists`` provides an information about error type; otherwise ``exists`` indicated whether
+the element exists.
```
What errors  can occur here?

mike1808 replied:
You are right. Particularly this method has very little chance to fail (or maybe doesn't have at all).

immerrr commented on docs/scripting-element-object.rst:
```diff
@@ -0,0 +1,393 @@
+Check whether the element exists in DOM. If the element doesn't exist almost all other
+methods will fail.
+
+**Signature:** ``ok, exists = element:exists()``
```
The flag is still here :)

kmike commented on docs/scripting-element-object.rst:
```diff
@@ -0,0 +1,393 @@
+Element Object
+==============
+
+Element objects are instanced by :ref:`splash-select`;
```
instantiated, created

kmike commented on docs/scripting-element-object.rst:
```diff
@@ -0,0 +1,393 @@
+.. note::
+
+    If the specified method returns another HTML element it will be represented as a table with
+    ``type`` property which equals to ``'node'``
```
why not Element class?

mike1808 replied:
It's already fixed in the other PR. I am going to refactor this PR.

kmike replied:
👍

kmike commented on docs/scripting-element-object.rst:
```diff
@@ -0,0 +1,393 @@
+
+Trigger mouse click event on the top-left corner of the element.
+
+**Signature:** ``ok, reason = element:mouse_click()``
```
It'd be nice to support relative coordinates here, defaulting to 0,0

mike1808 replied:
Ok
What if that coordinate exceed the element width or height? Just ignore it and click on the other element or throw an error?

kmike replied:
I'd say click anyways, without throwing an error.

kmike commented on docs/scripting-element-object.rst:
```diff
@@ -0,0 +1,393 @@
+
+Return a screenshot of the element in PNG format
+
+**Signature:** ``ok, shot = element:png{width=nil, height=nil, scale_method='raster'}``
```
I used a similar function in a project, and it was really nice to be able to take a screenshot of an area which is slightly larger than element itself. What do you think about adding 'pad' argument - a number of pixels to add to the screenshot around the element? Maybe it can also be a table with {top, left, bottm, right} padding widths. I know, more features.. :)

In implementation it'd be nice to have padding function separate and public, so that people can use it e.g. with get_bounds.

immerrr replied:
What about making the "padding" relative coordinates, i.e. positive values mean normal html padding, negative values - pad the image "inside" of the element.

On a sidenote, i don't think it's necessary to keep "width" and "height"  parameters with the same, quite obscure meaning as in the api. We might want to seize the opportunity to think of better names or better ways to describe the use cases for these parameters, not necessarily within this project.

kmike replied:
> What about making the "padding" relative coordinates, i.e. positive values mean normal html padding, negative values - pad the image "inside" of the element.

yeah, that's what I meant, but haven't expressed clearly!

> On a sidenote, i don't think it's necessary to keep "width" and "height" parameters with the same, quite obscure meaning as in the api. We might want to seize the opportunity to think of better names or better ways to describe the use cases for these parameters, not necessarily within this project.

What do you have in mind? It is nice to be able to rescale element when rendering; do you have a better API proposal?

mike1808 replied:
Nice idea about the padding. What do think, what should we in the following cases:

- for positive pads, if the padded image is going out of the web page bounds should we throw error or just fill with some color (maybe user-specified) 
- for negative pads, if the padding image has no width or height at all: throw error or return an empty image?

mike1808 replied:
> What do you have in mind? It is nice to be able to rescale element when rendering; do you have a better API proposal?

What if we rename the `size` to `rescale_to`?

kmike replied:
> for positive pads, if the padded image is going out of the web page bounds should we throw error or just fill with some color (maybe user-specified)

+1 to fill with a color. I recall there was some filling in existing rendering methods; imho we should do the same here (/cc @immerrr)

> for negative pads, if the padding image has no width or height at all: throw error or return an empty image?

my vote is to return an empty image.

> What if we rename the size to rescale_to? Will it make easier to understand the meaning of that argument?

there is no 'size' argument, there are width and height. User can pass wither both of them, or only one of them. Also, in splash:png and splash:jpeg they work differently: width causes page to be rescaled, while height causes page to be truncated at the bottom. So rescale_to is not the same as width/height.

kmike commented on docs/scripting-element-object.rst:
```diff
@@ -0,0 +1,393 @@
+
+This methods do the following:
+
+* checks whether the selected element is editable
```
Why is this check needed?

kmike replied:
E.g. element may become editable only after a click, or it may listen to keyboard events without being editable

mike1808 replied:
The overall idea of "sending keys to element" is so obvious. What does it mean to send keys to the element? 

The original idea was taken from casperj's "setField" method which do almost the same:
- checks whether the element is editable 
- focuses on it 
- send key events

However, in Splash we have more control over the web page (e.g. we can simulate a "real" mouse click, w/o using JS). So, I think you point is good. :)

kmike commented on docs/scripting-ref.rst:
```diff
@@ -2436,3 +2436,21 @@ arguments passed to splash, `username` and `password`.
+
+* selector - valid CSS selector
+
+**Returns:** a :ref:`splash-element`.
```
Async flag is missing

mike1808 replied:
Should I write some small guide why this feature is useful?

kmike replied:
Yes, and more examples could be nice!

kmike commented on splash/browser_tab.py:
```diff
@@ -104,6 +106,11 @@ def _init_webpage(self, verbosity, network_manager, splash_proxy_factory, render
         self.web_view.resize(
             QSize(*map(int, defaults.VIEWPORT_SIZE.split('x'))))
 
+    def _init_elements_storage(self):
```
It'd be nice to have this work lazily, when a first HtmlElement instance is created. This way Splash won't be changing global scope unless required.

mike1808 replied:
Good point.

kmike commented on splash/browser_tab.py:
```diff
@@ -1095,6 +1121,16 @@ def log(self, message, min_level=None):
+
+    @pyqtSlot(name="getId", result=str)
+    def get_id(self):
+        return str(uuid.uuid1())
```
It seems I'm missing something - why is there two uuid.uuid1() calls, one to set self.name, and another in get_id()?

mike1808 replied:
`get_id` is used in JS for retrieving a unique ID for the element. However, it's a good idea to call the same function in the constructor.

kmike replied:
Ah, so self.name is an id of elements storage element, and get_id generates an id of an element inside this storage? This makes sense. Could you please add docstrings to make it more clear?

kmike commented on splash/casperjs_utils.py:
```diff
@@ -0,0 +1,114 @@
+# Copyright (c) 2011-2015 Nicolas Perriault
```
could you please add a comment with a link to the casperjs file with relevant source code?

kmike commented on splash/jsutils.py:
```diff
@@ -39,6 +40,15 @@ def escape_js(*args):
+            else if (o instanceof ClientRect) {
+                return {
+                    top: o.top, left: o.left, bottom: o.bottom, right: o.right,
+                    width: o.width, height: o.height,
```
👍 

I wonder if it is possible to assign a non-float value to a ClientRect attribute in webkit used in Splash, bypassing this sanitizing function

mike1808 replied:
Why do you want to assign value to the client rect? It won't affect on anything.

kmike replied:
Someone may want to assign a value to client rect in order to bypass sanitanization.

kmike commented on splash/qtrender_lua.py:
```diff
@@ -80,7 +80,7 @@ def get_last(self):
 
 def command(table_argument=False, sets_callback=False,
-            decode_arguments=True):
+            decode_arguments=True, flag=False):
```
could you please use a longer name for `flag` variable? e.g. `errors_as_flags` or something like this (I'm not exactly sure about its meaning)?

kmike commented on splash/tests/test_execute.py:
```diff
@@ -3925,3 +3925,433 @@ def test_key_error(self):
+            local p = splash:select('p')
+            local form = splash:select('form#login')
+            local username = splash:select('input[name="username"]')
+            local password = splash:select('input[name="password"]')
```
From this example it looks like element:select() is a good idea - I'd better write `form:select(...)` here, as ideally the lookup should be limited to `form#login`.

mike1808 replied:
In my PR #490 you can do "form.node:querySelector()"

kmike replied:
hm, by the way, is there a difference between `node:querySelector(css)` and `:select(css)`? Can we make `:select(css)` a shortcut for `.node:querySelector(css)`? Are there extra features `:select()` provides?

kmike commented on splash/tests/test_execute.py:
```diff
@@ -3925,3 +3925,433 @@ def test_key_error(self):
+
+            local p = splash:select('p')
+            local ok, parent = p:node_property('parentNode')
+            local ok, remove = parent:node_method('remove')
```
`splash:select('p'):node_property('parentNone'):node_method('remove')()` is so much nicer when you're writing quick scraping scripts :) I woder if we should sacrifice `ok, result` consistency and start returning nil objects or raising errors

immerrr replied:
There's a branch that would allow `splash:select('p').parentNode:remove()`, that raises errors if something goes wrong, and I have high hopes for it. +1 from me.

Also, while writing it made me wonder if we want to have `.parentNode` for attributes and `:remove` for methods.

kmike commented on splash/utils.py:
```diff
@@ -89,7 +89,7 @@ def get_alive():
         '_ExposedRequest', '_ExposedBoundRequest',
         '_ExposedResponse', '_ExposedBoundResponse',
-        '_ExposedTimer',
+        '_ExposedTimer', '_ExposedElement',
```
could you please add ElementsStorage here?

mike1808 commented:
@immerrr @kmike I've moved all the work to #490 because I've did quite much refactoring. I'm going to fix all your notes in that branch. And if it OK for you I'm going to close this PR.

kmike commented:
Makes sense 👍
